### PR TITLE
Maintenance: increase portability of cf_gen automake rules

### PR DESCRIPTION
**Pull request by kinkie**
The MacOS linker uses a subdirectory named [target].dSYM to hold debug
information. Ensure it is removed by 'make clean'.

### CODE REVIEW

File: src/Makefile.am [hunk 1]
```diff
@@ -707,6 +707,9 @@ uninstall-local: squid.conf.default
 	@$(SHELL) $(top_srcdir)/scripts/remove-cfg.sh "$(RM)" $(DESTDIR)$(DEFAULT_MIME_TABLE) $(srcdir)/mime.conf.default
 	@$(SHELL) $(top_srcdir)/scripts/remove-cfg.sh "$(RM)" $(DESTDIR)$(DEFAULT_CONFIG_FILE) squid.conf.default
 
+distclean-local:
```
Comment (rousskov) [inline src/Makefile.am:710]: FWIW, I could not quickly figure out whether .dSYM should be deleted during "make clean" or "make distclean". Are you sure that "make clean" is not the right target for this?

File: src/Makefile.am [hunk 2]
```diff
@@ -707,6 +707,9 @@ uninstall-local: squid.conf.default
...
+	test -d "cf_gen.dSYM" && rm -rf "cf_gen.dSYM" || $(TRUE)
```
Comment (rousskov) [inline src/Makefile.am:711]: > The cf_gen target may leak this subdirectory for cf_gen

What makes cf_gen special? In other words, why do we need to add this special cleanup for cf_gen but not for other binaries? Perhaps the problem should be addressed by modifying how cf_gen is built?

**Review by rousskov — CHANGES_REQUESTED** [review]
I committed a polishing fix; please check.

I am not requesting any changes at this point, just more info.

File: src/Makefile.am (hunk 2 above)
Comment (kinkie) [reply src/Makefile.am:711]: I think it could be be because it's the only noinst_PROGRAM we are buildingq

Comment (rousskov) [reply src/Makefile.am:711]: Does temporary making cf_gen an installed program solves the problem (without this PR changes)?

File: src/Makefile.am (hunk 1 above)
Comment (yadij) [reply src/Makefile.am:710]: AIUI, if an object is generated by make `clean` should erase it. `distclean` is for the extra bits that can be re-generated by another build tool, eg re-running `configure`.

File: src/Makefile.am (hunk 2 above)
Comment (kinkie) [reply src/Makefile.am:711]: > Does temporary making cf_gen an installed program solves the problem (without this PR changes)?

Even if it did pass the build test, it would make no sense to keep it in, because by doing so it would cause cf_gen to be installed to a target system. The only reason to do it would be to possibly file a bug against autoconf

File: src/Makefile.am (hunk 1 above)
Comment (kinkie) [reply src/Makefile.am:710]: yes, clean-local works. Updating

File: src/Makefile.am (hunk 2 above)
Comment (kinkie) [reply src/Makefile.am:711]: Okay, I figured it out.
TL;DR: we override too much of autotools to build cf_gen. Updates coming up

**Review by rousskov — APPROVED** [review]
I trust you have tested that this change did not affect what files are archived by "make dist".

**Comment by rousskov** [comment]
I polished PR description a little. I do not insist on those changes.

File: src/Makefile.am [hunk 3]
```diff
@@ -624,10 +622,6 @@ test_cache_digest: test_cache_digest.o CacheDigest.o debug.o globals.o store_key
 ## If autodependency works well this is not needed anymore
 cache_cf.o: cf_parser.cci
 
-# cf_gen builds the configuration files.
-cf_gen$(EXEEXT): $(cf_gen_SOURCES) $(cf_gen_DEPENDENCIES) cf_gen_defines.cci
```
Comment (yadij) [inline src/Makefile.am:628]: Removal of this rule causes cross-compile to break with "cf_gen.exe: cannot execute binary file: Exec format error"

**Review by yadij — CHANGES_REQUESTED** [review]
This breaks build for me

File: src/Makefile.am (hunk 3 above)
Comment (kinkie) [reply src/Makefile.am]: reverted

File: src/Makefile.am [hunk 4]
```diff
@@ -156,11 +156,9 @@ bin_PROGRAMS =
 libexec_PROGRAMS = \
 	$(UNLINKD)
 
-cf_gen_SOURCES = cf_gen.cc
-nodist_cf_gen_HEADER = cf_gen_defines.cci
+cf_gen_SOURCES = cf_gen.cc cf_gen_defines.cci
```
[the remark this answers is not in the document]
Comment (kinkie) [reply src/Makefile.am]: yes. Reverted.

File: src/Makefile.am [hunk 5]
```diff
@@ -707,6 +707,9 @@ uninstall-local: squid.conf.default
 	@$(SHELL) $(top_srcdir)/scripts/remove-cfg.sh "$(RM)" $(DESTDIR)$(DEFAULT_MIME_TABLE) $(srcdir)/mime.conf.default
 	@$(SHELL) $(top_srcdir)/scripts/remove-cfg.sh "$(RM)" $(DESTDIR)$(DEFAULT_CONFIG_FILE) squid.conf.default
 
+clean-local:
+	test -d "cf_gen.dSYM" && rm -rf "cf_gen.dSYM" || $(TRUE)
```
Comment (rousskov) [inline src/Makefile.am:711]: Automake [documentation](https://www.gnu.org/software/automake/manual/html_node/Clean.html) suggests a simpler variant that uses a recipe line prefix to [handle errors](https://www.gnu.org/software/make/manual/make.html#Errors-in-Recipes):

```suggestion
	-rm -rf cf_gen.dSYM
```

There is also no need to "quote" cf_gen.dSYM AFAICT (the above suggestion remove quotes).

I have not tested the above. If this simpler version works, we should use it IMO.

N.B. Squid already uses `-` recipe line prefixes in contrib/Makefile.am (at least).

**Review by rousskov — CHANGES_REQUESTED** [review]

File: src/Makefile.am (hunk 5 above)
Comment (kinkie) [reply src/Makefile.am:711]: The theoretical difference is that the existing line will not try to remove the directory if it's not there, where the proposed change will remove it unconditionally but then not fail if it can't (e.g. because it doesn't exist).

I'm not attached to the existing version and will accept the change to allow the PR to continue, but I doubt it's worth a blocking change request

**Review by yadij — APPROVED** [review]
LGTM. I have adjusted the description again to document what the current change is.

File: src/Makefile.am (hunk 5 above)
Comment (rousskov) [reply src/Makefile.am]: > The theoretical difference is that the existing line will not try to remove the directory if it's not there, where the proposed change will remove it unconditionally but then not fail if it can't (e.g. because it doesn't exist).

Not exactly: AFAICT, `rm` does check whether the file is there (because it wants to get filesystem information about the file, very similar to what `test -d` would have to do), and `rm -rf` does not fail when the directory does not exist (it exits with zero status code instead).

There are, however, other, more important (IMO) differences. The proposed version is a lot simpler and does not introduce a yet another way of deleting things in Squid Automake files.
 
> I'm not attached to the existing version and will accept the change to allow the PR to continue, but I doubt it's worth a blocking change request

Thank you for accepting this change.

**Review by rousskov — APPROVED** [review]

**Comment by squidadm** [comment]
queued for backport to v6
